rpc/gnmi/gnmi-config-subscriptions.md

```diff
@@ -0,0 +1,187 @@
+target to indicate a commit boundary to the client.
+
+## 2.1 Proto
+
+The extension contains a message called `ConfigSubscription` that carries
+one of 2 types of actions. `ConfigSubscriptionStart` or
+`ConfigSubscriptionSyncDone`
+
+```proto
+// ConfigSubscription extension allows clients to subscribe to configuration
+// schema nodes only.
+message ConfigSubscription {
+  oneof action {
+    // ConfigSubscriptionStart is sent by the client in the SubscribeRequest
+    ConfigSubscriptionStart start = 1;
+    // ConfigSubscriptionSyncDone is sent by the server in the SubscribeResponse
+    ConfigSubscriptionSyncDone sync_done = 2;
+  }
+}
+
+// ConfigSubscriptionStart is used to indicate to a target that for a given set
+// of paths in the SubscribeRequest, the client wishes to receive updates
+// for the configuration schema nodes only.
+message ConfigSubscriptionStart {}
+
+// ConfigSubscriptionSyncDone is sent by the server in the SubscribeResponse
+// after all the updates for the configuration schema nodes have been sent.
+message ConfigSubscriptionSyncDone {
+  // ID of a commit confirm operation as assigned by the client
+  // see Commit Confirm extension for more details.
+  string commit_confirm_id = 1;
+  // ID of a commit as might be assigned by the server
+  // when registering a commit operation.
+  string server_commit_id = 2;
+}
+```
+
+## 2.2 Actions
+
+### 2.2.1 ConfigSubscriptionStart
+
+A `ConfigSubscriptionStart` message is used by a gNMI client in a
+`SubscribeRequest` to indicate that it wants to start a ConfigSubscription.
+The target must respond exclusively with configuration data relevant to the
+created subscription.
+The target must respond with an initial set of updates followed by a
+`SubscribeResponse` where the `sync_update` field is set to `true`.
+However, if the `updates_only` field in the `SubscribeRequest` was set to
+`true`, the target should skip sending the initial updates and only send
+a `SubscribeResponse` with `sync_update` set to `true` as per the base gNMI
+specification.
+
+### 2.2.2 ConfigSubscriptionSyncDone
+
+A `ConfigSubscriptionSyncDone` message is used by a gNMI target in a
+`SubscribeResponse` to indicate a commit boundary to the client.
+A commit boundary marks the completion point of a specific set of
+configuration changes.
+It indicates that all changes within that set have been committed and
+that all notifications triggered by the commit have been streamed to
+the client.
+
+The `ConfigSubscriptionSyncDone` message includes two optional fields:
+
+* `commit_confirm_id`: A commit confirm ID assigned by the client which
+initiated the commit.
+The commit can be initiated via gNMI (using the CommitConfirmed Extension),
+NETCONF, or any other management interface. Applicable only if the commit
+confirmed option is used.
+* `server_commit_id`: An optional internal ID assigned by the target.
+
+In the case a commit happens before the `sync_response: true` and the server
```

Comment:
> is the implication here that we should hold back this sync_reponse: true until all the leaves that are the result of a new commit are sent?

It means that we should hold back the `ConfigSubscriptionSyncDone` until after `sync_reponse: true`.
Essentially, do not send any `ConfigSubscriptionSyncDone` until the initial updates are sent, even if all updates related to a commit were already streamed to the collector.

Example:

- A clientA creates a subscription with this extension to root `/`
- While the initial updates are still being streamed to clientA, clientB changes the hostname from `dut1` to `dut2`
- ClientA might receive the hostname twice (`dut1` then `dut2`) but would only receive a `ConfigSubscriptionSyncDone` after all the initial updates are streamed and a `sync_reponse: true` is sent.

Reply:
this would be worth clarifying in the text -- since (as above) I read it the other way around. I think simply saying that `ConfigSubscriptionSyncDone` cannot be sent until `sync_response: true` has been sent is reasonable. (this will still need some per-client state to be stored about whether the initial sync has been completed -- but at least does not mean that the `sync_response` behaviour is changed.)

Reply:
The paragraph is more about allowing a server to send the committed changes updates before the `sync_response: true`. Meaning that updates do not have to be buffered but the `ConfigSubscriptionSyncDone` has to be.
Rephrasing to this:

```
In the case a commit happens before the `sync_response: true` the server
cannot send a `ConfigSubscriptionSyncDone` until the `sync_response: true` has
been sent. The server may send the committed changes updates before the
`sync_response: true`.
```

Reply:
Thanks, LGTM.